### PR TITLE
Improve labels map computation - WIP

### PR DESCRIPTION
**Pull request by frheault**
Big refactor of labels maps computation, this would require to put this script almost as the first (after centroids/uniformization) in the tractometry nextflow pipeline.

This script now removes isolated voxels when there is only one streamline in it with no neighbor. It then cut streamlines to keep as many as possible. Making the bundle less noisy at the edge without actually losing fanning.

Since the centroids are aligned with a whole brain transformation, a tiny SLR was added to center the centroids onto the core of the bundle, this seems to help a lot with the alignment. The transformation is mostly rigid only, so shorter/cut bundle due to pathology are not mixing up labels by shortening the centroid. I allow for a 0.05 scaling in each direction.

Then it compute the labels maps the same way as before, but then it finds streamline portions that jumps too many labels (nb_division // 5) and get the longest 'continuous' sequences of labels and chop out the rest which cleans out the edge of nearly looping bundles (AF, UF, CC) where confusion occurs due to the simplistic euclidian distance.

The assignment of labels is using the euclidian distance because that's what is intuitive visually, that's how we expect the labeling to work. However, the distance is now computed using an approximation of the distance to the centroids following the WM only. By iteratively dilating the centroids mask (and restraining it) I count the number of iteration to get to each voxel. This allows to spot voxels where you have to follow a lot of curves to get to rather than using the euclidian distance that goes over the anatomy without knowing it.

### CODE REVIEW

**Review by mdesco — COMMENTED** [review]
Small -h corrections to do.

Aside from this, wondering about labels 1 and N. If low N, works fine. For high N, the label map 1 and N are "huge" with respect to the others. See images below. 

Comments? Ideas to minimize this "bias"? 


AF_100
AF_50
AF_5
5pts
50pts
100pts

File: scripts/scil_compute_bundle_voxel_label_map.py [hunk 1]
```diff
@@ -33,27 +43,95 @@ def _build_arg_parser():
...
+    p.add_argument('out_labels_map',
                    help='Nifti image with corresponding labels.')
-    p.add_argument('--upsample', type=float, default=1,
-                   help='Upsample reference grid by this factor. '
-                        '[%(default)s]')
+    p.add_argument('out_distances_map',
+                   help='Nifti image showing distances to centroids.')
+
+    p.add_argument('--nb_pts', type=int,
+                   help='Number of divisions for the bundles.')
```
[the remark this answers is not in the document]
Comment (frheault) [reply scripts/scil_compute_bundle_voxel_label_map.py:52]: Yes, the default is not to touch the nb_of points in the centroids. I will add a sentence to make it clear.

**Comment by frheault** [comment]
@mdesco All centroids are shorter than the actual bundles. Mine or if you compute it from the data, are a few millimeters 10-20mm shorter on both ends since no streamline actually run the whole length. So if you do 5-10 subdivisions it won't be noticeable that the first/last labels are longer.

I tried to make it longer but then the opposite happens, I end up having a lot of bundles with missing labels 1 to 5 and N-5 to N and it is also counter-intuitive. Not sure which one is the best, what do you think?

**Comment by mdesco** [comment]
I think what you do is fine but we have to know and tell our users of this behaviour.
1. Most likely labels 1 and N to be excluded from the stats if N is high
2. If endpoints stats are wanted, user can use the "true" endpoints_map xlsx, no?

I would leave it as is if it is your preferred solution. We will adjust to it

**Comment by frheault** [comment]
That's a solution, but 1 and N are just much bigger. So if the person takes into account the volume and the STD it could be "fine" for the analysis too. If it is for an analysis like functional stuff (Fred) and its assume a similar length, yes they would be discarded. My bet is that above 20 it would be a good idea to QA the size in case (it is bundles dependant so I cant say)

Endpoints maps are actually spread across all labels, not just 1 and N anyway. So that's would be a solution to a different problem I think

**Review by mdesco — APPROVED** [review]

File: scripts/scil_compute_bundle_profiles.py [hunk 2]
```diff
@@ -85,6 +85,7 @@ def main():
 
     # Centroid - will be use as reference to reorient each streamline
     if args.in_centroid:
+        # TODO no test here ...
```
[the remark this answers is not in the document]
Comment (frheault) [reply scripts/scil_compute_bundle_profiles.py:88]: Should we remove this script, its like the easier the call equivalent of what we do in Tractometry.
It does almost the same thing except it is not really the same.

I think its weird to have this 'analogous' script that is not used in Tractometry, people got confused in the past. Everytime I see it I am unsure of how to use it and how it works compare to scil_compute_mean_std_per_point.py

File: scilpy/tractanalysis/tools.py [hunk 3]
```diff
@@ -158,3 +160,51 @@ def compute_connectivity(indices, atlas_data, real_labels, segmenting_func):
...
+                if entry_found:
+                    entry_found = False
+                    if curr_len > longest_seq[1] - longest_seq[0]:
+                        longest_seq = (last_success, ind-1)
+                        curr_len = 0
+        if longest_seq[1] != 0:
+            new_streamlines.append(streamline[longest_seq[0]:longest_seq[1]])
+
+    new_sft = StatefulTractogram.from_sft(new_streamlines, sft)
+    return filter_streamlines_by_length(new_sft, min_length=5, max_length=300)
```
Comment (arnaudbore) [inline scilpy/tractanalysis/tools.py:210]: Maybe you should take the min length and max length from the original tractogram.

**Review by arnaudbore — APPROVED** [review]
Appart from the delete or nor delete compute_bundle_profiles LGTM.

**Pull request merged**